File: benchmark/synduce/list/zeros_ones.f
Inductive List = elt Bool | cons {Bool, List};
Inductive CList = single Bool | concat {CList, CList};

cat_list = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    elt a -> cons {a, ys}
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    single a -> elt a
  | concat {a, b} -> cat_list (f a) (f b)
  end
);

spec = \xs: List. (fix (
  \f: List -> {Bool, Bool, Bool}. \xs: List.
  match xs with
    elt a -> {a, true, a}
  | cons {hd, tl} -> 
    let result = f tl in
    let new_an = and result.1 hd in
    let new_bn = and result.2 (or result.1 (not hd)) in
    {new_an, new_bn, result.3}
  end
) xs).2;

main = \xs: CList. spec (repr xs);

File: ml-benchmark/synduce/list/zeros_ones.ml
type list_ = Elt of bool | Cons of bool * list_
type clist = Single of bool | Concat of clist * clist

let rec cat_list xs ys =
  match xs with
  | Elt a -> Cons (a, ys)
  | Cons (a, b) -> Cons (a, cat_list b ys)

let rec repr xs =
  match xs with
  | Single a -> Elt a
  | Concat (a, b) -> cat_list (repr a) (repr b)

let spec xs =
  let rec f xs =
    match xs with
    | Elt a -> (a, true, a)
    | Cons (hd, tl) ->
      let result = f tl in
      match result with
      | (r1, r2, r3) ->
        let new_an = r1 && hd in
        let new_bn = r2 && (r1 || not hd) in
        (new_an, new_bn, r3)
  in
  match f xs with
  | (_, b, _) -> b

let program xs = spec (repr xs)